File: src/matrix.cpp
#include "matrix.h"
#include "comparinator.h"
#include "tuple.h"
#include <cmath>
#include <iostream>

Matrix::Matrix()
{
	mbrRows = 4;
	mbrColumns = 4;
	this->mbrGrid = new double* [mbrRows];
	for (int i = 0; i < mbrRows; i++)
	{
		this->mbrGrid[i] = new double[mbrColumns];
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = 0;
		}
	}
}
Matrix::Matrix(int rows, int columns)
{
	mbrRows = rows;
	mbrColumns = columns;
	this->mbrGrid = new double* [mbrRows];
	for (int i = 0; i < mbrRows; i++)
	{
		this->mbrGrid[i] = new double[mbrColumns];
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = 0;
		}
	}
}

Matrix::Matrix(const Matrix& other)
{
	//std::cout << "matrix copy ctor" << std::endl;
	this->mbrRows = other.mbrRows;
	this->mbrColumns = other.mbrColumns;
	//std::cout << "this: " << mbrRows << mbrColumns << " other:" << other.mbrRows << other.mbrColumns << std::endl;
	this->mbrGrid = new double* [mbrRows];
	for (int i = 0; i < mbrRows; i++)
	{
		this->mbrGrid[i] = new double[mbrColumns];
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = other.mbrGrid[i][j];
			//std::cout << "this: " << mbrGrid[i][j] << " other:" << other.mbrGrid[i][j] << std::endl;
		}
	}
}

Matrix::Matrix(int rows, int columns, double* values)
{
	mbrRows = rows;
	mbrColumns = columns;
	this->mbrGrid = new double* [mbrRows];
	for (int i = 0; i < mbrRows; i++)
	{
		this->mbrGrid[i] = new double[mbrColumns];
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = values[(i * columns) + j];
		}
	}
}

Matrix::~Matrix()
{
	//std::cout << "deleting matrix" << &grid << std::endl;
	for (int i = 0; i < mbrRows; i++)
	{
		delete[] this->mbrGrid[i];
		this->mbrGrid[i] = nullptr;
	}
	delete[] this->mbrGrid;
	this->mbrGrid = nullptr;
}

Matrix& Matrix::operator=(const Matrix other)
{
	if (this == &other) return *this;
	//std::cout << "matrix copy assign" << std::endl;
	this->mbrRows = other.mbrRows;
	this->mbrColumns = other.mbrColumns;
	for (int i = 0; i < mbrRows; i++)
	{
		delete[] this->mbrGrid[i];
		this->mbrGrid[i] = nullptr;
	}
	delete[] this->mbrGrid;
	this->mbrGrid = nullptr;
	this->mbrGrid = new double* [mbrRows];
	for (int i = 0; i < mbrRows; i++)
	{
		this->mbrGrid[i] = new double[mbrColumns];
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = other.mbrGrid[i][j];
		}
	}
	return *this;
}

bool Matrix::operator==(const Matrix other) const
{
	//std::cout << "matrix equality operator" << std::endl;
	if (other.mbrRows != mbrRows || other.mbrColumns != mbrColumns)
	{
		return false;
	}
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			//std::cout << "addresses this:" << &grid[i][j] << ":other:" << &other.grid[i][j] << std::endl;
			//std::cout << "values this:" << grid[i][j] << ":other:" << other.grid[i][j] << std::endl;
			if (other.mbrGrid[i][j] != this->mbrGrid[i][j])
			{
				return false;
			}
		}
	}
	return true;
}

bool Matrix::checkEqual(const Matrix other)
{
	//std::cout << "matrix equality" << std::endl;
	if (other.mbrRows != mbrRows || other.mbrColumns != mbrColumns)
	{
		return false;
	}
	Comparinator ce = Comparinator();
	//std::cout << "row == column" << std::endl;
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			//std::cout << "addresses this:" << & grid[i][j] << ":other:" << &other.grid[i][j] << std::endl;
			//std::cout << "values this:" << grid[i][j] << ":other:" << other.grid[i][j] << std::endl;
			if (!ce.checkFloat(other.mbrGrid[i][j],this->mbrGrid[i][j]))
			{
				return false;
			}
		}
	}
	return true;
}

Matrix* Matrix::operator*(const Matrix other)
{
	//std::cout << "matrix multiplication" << std::endl;
	Matrix* result = new Matrix(other.mbrRows, other.mbrColumns);
	for (int i = 0; i < result->mbrRows; i++)
	{
		for (int j = 0; j < result->mbrColumns; j++)
		{
			double value = 0;
			for (int k = 0; k < result->mbrColumns; k++)
			{
				//std::cout << "i:" << i << "j:" << j << "k:" << k << "=" << grid[i][k] << "*" << other.grid[k][j] << std::endl;
				value += mbrGrid[i][k] * other.mbrGrid[k][j];
			}
			//std::cout << value << std::endl;
			result->setRC(i, j, value);
		}
	}
	return result;
}

Tuple Matrix::operator*(Tuple const other)
{
	if (mbrRows != 4 && mbrColumns != 4) return other;
	//std::cout << "matrix tuple multiplication" << std::endl;
	double pseudoMatrix[4] = {other.argX, other.argY, other.argZ, other.argW};
	double results[4] = { 0,0,0,0 };
	double res = 0;
	for (int tuple = 0; tuple < 4; tuple++)
	{
		res = 0;
		for (int col = 0; col < mbrColumns; col++)
		{
			//std::cout << "i:" << col << "j:" << tuple << "k:" << res << "=" << pseudoMatrix[col] << "*" << grid[tuple][col] << std::endl;
			res += mbrGrid[tuple][col] * pseudoMatrix[col];
		}
		results[tuple] = res;
	}
	return Tuple(results[0],results[1],results[2],results[3]);
}
Point Matrix::operator*(const Point other)
{
	Tuple tp = Tuple(other.argX, other.argY, other.argZ, other.argW);
	Tuple res = *(this) * tp;
	return Point(res.argX, res.argY, res.argZ);
}
Vector Matrix::operator*(const Vector other)
{
	Tuple tp = Tuple(other.argX, other.argY, other.argZ, other.argW);
	Tuple res = *(this) * tp;
	return Vector(res.argX, res.argY, res.argZ);
}

bool Matrix::checkInvertible() {
	return determinant() != 0;
}
bool Matrix::checkValid(int row, int column)
{
	return row >= 0 && row < mbrRows && column >= 0 && column < mbrColumns;
}

double Matrix::getRC(int row, int column)
{
	if (checkValid(row, column))
	{
		return this->mbrGrid[row][column];
	}
	return 0;
}

void Matrix::setRC(int row, int column, double value)
{
	if (checkValid(row, column))
	{
		this->mbrGrid[row][column] = value;
	}
}

Matrix* Matrix::transpose()
{
	Matrix* copy = new Matrix(mbrColumns, mbrRows);
	for (int i = 0; i < copy->mbrRows; i++)
	{
		for (int j = 0; j < copy->mbrColumns; j++)
		{
			copy->setRC(i, j, this->getRC(j, i));
			//std::cout << "i:" << i << "j:" << j << "now: " << this->getRC(j, i) << "=" << copy->getRC(i, j) << std::endl;
		};
	};
	return copy;
}

double Matrix::determinant()
{
	double determinant = 0;
	if (mbrRows == 2 && mbrColumns == 2) determinant = mbrGrid[0][0] * mbrGrid[1][1] - mbrGrid[0][1] * mbrGrid[1][0];
	else
	{
		for (int i = 0; i < mbrColumns; i++)
		{
			determinant += cofactor(0, i) * getRC(0, i);
		};
	};
	return determinant;
}

Matrix* Matrix::submatrix(int row, int column)
{
	Matrix* sub = new Matrix(mbrRows - 1, mbrColumns - 1);
	int subrow = 0;
	int subcol = 0;
	int i = 0;
	int j = 0;
	while (i < mbrRows)
	{
		if (i == row) { i += 1; }
		j = 0;
		subcol = 0;
		while (j < mbrColumns)
		{
			if (j == column) { j += 1; }
			sub->setRC(subrow, subcol, this->getRC(i, j));
			//std::cout << "i:" << subrow << "j:" << subcol << "now: " << this->getRC(i, j) << "=" << sub->getRC(subrow, subcol) << std::endl;
			j += 1;
			subcol += 1;
		}
		i += 1;
		subrow += 1;
	}
	return sub;
}

double Matrix::minor(int row, int column)
{
	return this->submatrix(row, column)->determinant();
}

double Matrix::cofactor(int row, int column)
{
	return (row + column) % 2 == 0 ? this->minor(row, column) : -1 * this->minor(row, column);
}

Matrix* Matrix::invert()
{
	Matrix* inverse = new Matrix(mbrRows, mbrColumns);
	if (!checkInvertible()) return inverse;
	double dm = this->determinant();
	for (int i = 0; i < this->mbrRows; i++)
	{
		for (int j = 0; j < this->mbrColumns; j++)
		{
			double cf = this->cofactor(i, j);
			inverse->setRC(j, i, cf / dm);
			//std::cout << "j:" << j << "i:" << i << "now: " << cf << "/" << dm << "=" << inverse->getRC(j, i) << std::endl;
		};
	};
	return inverse;
}
Matrix* Matrix::identity()
{
	Matrix* identity = new Matrix(mbrRows, mbrColumns);
	for (int i = 0; i < identity->mbrRows; i++)
	{
		for (int j = 0; j < identity->mbrColumns; j++)
		{
			i == j ? identity->mbrGrid[i][j] = 1 : identity->mbrGrid[i][j] = 0;
		};
	};
	return identity;
}
Matrix* Matrix::rotateX(double radians)
{
	XRotationMatrix rotate = XRotationMatrix(radians);
	return *this * rotate;
}
Matrix* Matrix::rotateY(double radians)
{
	YRotationMatrix rotate = YRotationMatrix(radians);
	return *this * rotate;
}
Matrix* Matrix::rotateZ(double radians)
{
	ZRotationMatrix rotate = ZRotationMatrix(radians);
	return *this * rotate;
}
Matrix* Matrix::scale(int x, int y, int z)
{
	ScalingMatrix scale = ScalingMatrix(x,y,z);
	return *this * scale;
}
Matrix* Matrix::translate(int x, int y, int z)
{
	TranslationMatrix translate = TranslationMatrix(x,y,z);
	return *this * translate;
}
Matrix* Matrix::shear(double xy, double xz, double yx, double yz, double zx, double zy)
{
	ShearingMatrix shear = ShearingMatrix(xy, xz, yx, yz, zx, zy);
	return *this * shear;
}

IdentityMatrix::IdentityMatrix (int rows, int columns) : Matrix(rows = rows, columns = columns)
{
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			i == j ? this->mbrGrid[i][j] = 1 : this->mbrGrid[i][j] = 0;
		};
	};
}

TranslationMatrix::TranslationMatrix(double x, double y, double z) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = 1;
	this->mbrGrid[0][3] = x;
	this->mbrGrid[1][3] = y;
	this->mbrGrid[1][1] = 1;
	this->mbrGrid[2][3] = z;
	this->mbrGrid[2][2] = 1;
	this->mbrGrid[3][3] = 1;
}

ScalingMatrix::ScalingMatrix(double x, double y, double z) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = x;
	this->mbrGrid[1][1] = y;
	this->mbrGrid[2][2] = z;
	this->mbrGrid[3][3] = 1;
}

XRotationMatrix::XRotationMatrix(double radians) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = 1;
	this->mbrGrid[1][1] = cos(radians);
	this->mbrGrid[1][2] = -sin(radians);
	this->mbrGrid[2][1] = sin(radians);
	this->mbrGrid[2][2] = cos(radians);
	this->mbrGrid[3][3] = 1;
}

YRotationMatrix::YRotationMatrix(double radians) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = cos(radians);
	this->mbrGrid[0][2] = sin(radians);
	this->mbrGrid[1][1] = 1;
	this->mbrGrid[2][0] = -sin(radians);
	this->mbrGrid[2][2] = cos(radians);
	this->mbrGrid[3][3] = 1;
}

ZRotationMatrix::ZRotationMatrix(double radians) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = cos(radians);
	this->mbrGrid[0][1] = -sin(radians);
	this->mbrGrid[1][0] = sin(radians);
	this->mbrGrid[1][1] = cos(radians);
	this->mbrGrid[2][2] = 1;
	this->mbrGrid[3][3] = 1;
}

ShearingMatrix::ShearingMatrix(double xy, double xz, double yx, double yz, double zx, double zy) : Matrix(4, 4)
{
	this->mbrGrid[0][0] = 1;
	this->mbrGrid[0][1] = xy;
	this->mbrGrid[0][2] = xz;
	this->mbrGrid[1][0] = yx;
	this->mbrGrid[1][1] = 1;
	this->mbrGrid[1][2] = yz;
	this->mbrGrid[2][0] = zx;
	this->mbrGrid[2][1] = zy;
	this->mbrGrid[2][2] = 1;
	this->mbrGrid[3][3] = 1;
}

ViewMatrix::ViewMatrix() : Matrix()
{
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			i == j ? this->mbrGrid[i][j] = 1 : this->mbrGrid[i][j] = 0;
		};
	};
}

ViewMatrix::ViewMatrix(Point start, Point end, Vector up) : Matrix()
{
	Vector varForward = (end - start).normalize();
	Vector varLeft = varForward.cross(up.normalize());
	Vector varUp = varLeft.cross(varForward);
	double varValues[] = {varLeft.argX, varLeft.argY, varLeft.argZ, 0, varUp.argX, varUp.argY, varUp.argZ, 0, -varForward.argX, -varForward.argY, -varForward.argZ, 0, 0, 0, 0, 1};
	Matrix varOrientation = Matrix(4,4, varValues);
	Matrix* varResult = varOrientation * TranslationMatrix(-start.argX, -start.argY, -start.argZ);
	for (int i = 0; i < mbrRows; i++)
	{
		for (int j = 0; j < mbrColumns; j++)
		{
			this->mbrGrid[i][j] = varResult->mbrGrid[i][j];
			// std::cout << "addresses this:" << & grid[i][j] << ":other:" << &varResult->grid[i][j] << std::endl;
			// std::cout << "values this:" << grid[i][j] << ":other:" << varResult->grid[i][j] << std::endl;
		};
	};
}

File: src/Matrix.cs
using LibTuple;
using LibComparinator;
namespace LibMatrix;

public class Matrix {
	public List<List<double>> _fieldGrid;
	public int _fieldRows;
	public int _fieldColumns;
	public Matrix()
	{
		_fieldRows = 4;
		_fieldColumns = 4;
		_fieldGrid = new List<List<double>>();
		for (int i = 0; i < _fieldRows; ++i)
		{
			_fieldGrid.Add(new List<double>());
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i].Add(i == j ? 1 : 0);
			}
		}
	}
	public Matrix(int argRows, int argColumns)
	{
		_fieldRows = argRows;
		_fieldColumns = argColumns;
		_fieldGrid = new List<List<double>>();
		for (int i = 0; i < _fieldRows; ++i)
		{
			_fieldGrid.Add(new List<double>());
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i].Add(i == j ? 1 : 0);
			}
		}
	}
	public Matrix(Matrix argOther)
	{
		_fieldRows = argOther._fieldRows;
		_fieldColumns = argOther._fieldColumns;
		_fieldGrid = new List<List<double>>();
		for (int i = 0; i < _fieldRows; ++i)
		{
			_fieldGrid.Add(new List<double>());
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i].Add(argOther._fieldGrid[i][j]);
			}
		}
	}
	public Matrix(int argRows, int argColumns, List<double> argValues)
	{
		_fieldRows = argRows;
		_fieldColumns = argColumns;
		_fieldGrid = new List<List<double>>();
		for (int i = 0; i < _fieldRows; ++i)
		{
			_fieldGrid.Add(new List<double>());
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i].Add(argValues[(i * argColumns) + j]);
			}
		}
	}
	// public static bool operator==(Matrix argSelf, Matrix argOther)
	// {
	// 	if (argOther._fieldRows != argSelf._fieldRows || argOther._fieldColumns != argSelf._fieldColumns)
	// 	{
	// 		return false;
	// 	}
	// 	for (int i = 0; i < argSelf._fieldRows; ++i)
	// 	{
	// 		for (int j = 0; j < argSelf._fieldColumns; ++j)
	// 		{
	// 			if (argOther._fieldGrid[i][j] != argSelf._fieldGrid[i][j])
	// 			{
	// 				return false;
	// 			}
	// 		}
	// 	}
	// 	return true;
	// }
	// public static bool operator!=(Matrix argSelf, Matrix other)
	// {
	// 	if (other._fieldRows != argSelf._fieldRows || other._fieldColumns != argSelf._fieldColumns)
	// 	{
	// 		return true;
	// 	}
	// 	for (int i = 0; i < argSelf._fieldRows; ++i)
	// 	{
	// 		for (int j = 0; j < argSelf._fieldColumns; ++j)
	// 		{
	// 			if (other._fieldGrid[i][j] != argSelf._fieldGrid[i][j])
	// 			{
	// 				return true;
	// 			}
	// 		}
	// 	}
	// 	return false;
	// }
	public static Matrix operator*(Matrix argSelf, Matrix argOther)
	{
		Matrix varResult = new Matrix(argOther._fieldRows, argOther._fieldColumns);
		for (int i = 0; i < varResult._fieldRows; ++i)
		{
			for (int j = 0; j < varResult._fieldColumns; ++j)
			{
				double value = 0;
				for (int k = 0; k < varResult._fieldColumns; k++)
				{
					value += argSelf._fieldGrid[i][k] * argOther._fieldGrid[k][j];
				}
				varResult.SetRC(i, j, value);
			}
		}
		return varResult;
	}
	public static SpaceTuple operator*(Matrix argSelf, SpaceTuple argOther)
	{
		if (argSelf._fieldRows != 4 && argSelf._fieldColumns != 4) return argOther;
		List<double> varPseudoMatrix = new List<double>{argOther._fieldX, argOther._fieldY, argOther._fieldZ, argOther._fieldW};
		List<double> varResults = new List<double>{ 0,0,0,0 };
		double res = 0;
		for (int tuple = 0; tuple < 4; tuple++)
		{
			res = 0;
			for (int col = 0; col < argSelf._fieldColumns; col++)
			{
				res += argSelf._fieldGrid[tuple][col] * varPseudoMatrix[col];
			}
			varResults[tuple] = res;
		}
		return new SpaceTuple(varResults[0],varResults[1],varResults[2],varResults[3]);
	}
	public static Point operator*(Matrix argSelf, Point argOther)
	{
		SpaceTuple varTuple = new SpaceTuple(argOther._fieldX, argOther._fieldY, argOther._fieldZ, argOther._fieldW);
		SpaceTuple varResult = argSelf * varTuple;
		return new Point(varResult._fieldX, varResult._fieldY, varResult._fieldZ);
	}
	public static Vector operator*(Matrix argSelf, Vector other)
	{
		SpaceTuple varTuple = new SpaceTuple(other._fieldX, other._fieldY, other._fieldZ, other._fieldW);
		SpaceTuple varResult = argSelf * varTuple;
		return new Vector(varResult._fieldX, varResult._fieldY, varResult._fieldZ);
	}
	public bool CheckInvertible() {
		return GetDeterminant() != 0;
	}
	public bool CheckValid(int argPositionRow, int argPositionColumn)
	{
		return argPositionRow >= 0 && argPositionRow < _fieldRows && argPositionColumn >= 0 && argPositionColumn < _fieldColumns;
	}
	public bool CheckEqual(Matrix argOther)
	{
		if (argOther._fieldRows != _fieldRows || argOther._fieldColumns != _fieldColumns)
		{
			return false;
		}
		Comparinator varComp = new Comparinator();
		for (int i = 0; i < _fieldRows; ++i)
		{
			for (int j = 0; j < _fieldColumns; ++j)
			{
				if (!varComp.CheckFloat(argOther._fieldGrid[i][j],_fieldGrid[i][j]))
				{
					return false;
				}
			}
		}
		return true;
	}
	public double GetRC(int argPositionRow, int argPositionColumn)
	{
		if (CheckValid(argPositionRow, argPositionColumn))
		{
			return _fieldGrid[argPositionRow][argPositionColumn];
		}
		return 0;
	}
	public void SetRC(int argPositionRow, int argPositionColumn, double argValue)
	{
		if (CheckValid(argPositionRow, argPositionColumn))
		{
			_fieldGrid[argPositionRow][argPositionColumn] = argValue;
		}
	}
	public double GetDeterminant()
	{
		double varDeterminant = 0;
		if (_fieldRows == 2 && _fieldColumns == 2) varDeterminant = _fieldGrid[0][0] * _fieldGrid[1][1] - _fieldGrid[0][1] * _fieldGrid[1][0];
		else
		{
			for (int i = 0; i < _fieldColumns; ++i)
			{
				varDeterminant += GetCofactor(0, i) * GetRC(0, i);
			};
		};
		return varDeterminant;
	}
	public double GetMinor(int argPositionRow, int argPositionColumn)
	{
		return GetSubMatrix(argPositionRow, argPositionColumn).GetDeterminant();
	}
	public double GetCofactor(int argPositionRow, int argPositionColumn)
	{
		return (argPositionRow + argPositionColumn) % 2 == 0 ? GetMinor(argPositionRow, argPositionColumn) : -1 * GetMinor(argPositionRow, argPositionColumn);
	}
	public virtual Matrix GetTranspose()
	{
		Matrix varTranspose = new Matrix(_fieldColumns, _fieldRows);
		for (int i = 0; i < varTranspose._fieldRows; ++i)
		{
			for (int j = 0; j < varTranspose._fieldColumns; ++j)
			{
				varTranspose.SetRC(i, j, GetRC(j, i));
			};
		};
		return varTranspose;
	}
	public virtual Matrix GetSubMatrix(int argRows, int argColumns)
	{
		Matrix varSubMatrix = new Matrix(_fieldRows - 1, _fieldColumns - 1);
		int varSubRow = 0;
		int varSubCol = 0;
		int i = 0;
		int j = 0;
		while (i < _fieldRows)
		{
			if (i == argRows) { i += 1; }
			j = 0;
			varSubCol = 0;
			while (j < _fieldColumns)
			{
				if (j == argColumns) { j += 1; }
				varSubMatrix.SetRC(varSubRow, varSubCol, GetRC(i, j));
				j += 1;
				varSubCol += 1;
			}
			i += 1;
			varSubRow += 1;
		}
		return varSubMatrix;
	}
	public virtual Matrix GetInverse()
	{
		Matrix varInverse = new Matrix(_fieldRows, _fieldColumns);
		if (!CheckInvertible()) return varInverse;
		double dm = GetDeterminant();
		for (int i = 0; i < _fieldRows; ++i)
		{
			for (int j = 0; j < _fieldColumns; ++j)
			{
				double cf = GetCofactor(i, j);
				varInverse.SetRC(j, i, cf / dm);
			};
		};
		return varInverse;
	}
	public virtual Matrix GetIdentity()
	{
		Matrix varIdentity = new Matrix(_fieldRows, _fieldColumns);
		for (int i = 0; i < varIdentity._fieldRows; ++i)
		{
			for (int j = 0; j < varIdentity._fieldColumns; ++j)
			{
				varIdentity._fieldGrid[i][j] = (i == j ? 1 : 0);
			};
		};
		return varIdentity;
	}
	public virtual Matrix GetRotateX(double argRadians)
	{
		XRotationMatrix rotate = new XRotationMatrix(argRadians);
		return this * rotate;
	}
	public virtual Matrix GetRotateY(double argRadians)
	{
		YRotationMatrix rotate = new YRotationMatrix(argRadians);
		return this * rotate;
	}
	public virtual Matrix GetRotateZ(double argRadians)
	{
		ZRotationMatrix rotate = new ZRotationMatrix(argRadians);
		return this * rotate;
	}
	public virtual Matrix GetScale(double x, double y, double z)
	{
		ScalingMatrix scale = new ScalingMatrix(x,y,z);
		return this * scale;
	}
	public virtual Matrix GetTranslate(double x, double y, double z)
	{
		TranslationMatrix translate = new TranslationMatrix(x,y,z);
		return this * translate;
	}
	public virtual Matrix GetShear(double xy, double xz, double yx, double yz, double zx, double zy)
	{
		ShearingMatrix shear = new ShearingMatrix(xy, xz, yx, yz, zx, zy);
		return this * shear;
	}
	public virtual void RenderConsole() 
	{
		Console.WriteLine("Matrix::renderConsole()");
		for (int i = 0; i < _fieldRows; ++i)
		{
			Console.Write("[");
			for (int j = 0; j < _fieldColumns; ++j)
			{
				Console.Write($"{_fieldGrid[i][j]}, ");
			};
			Console.WriteLine("]");
		};
	}
};
public class IdentityMatrix : Matrix {
	public IdentityMatrix(int rows, int columns) {
		for (int i = 0; i < _fieldRows; ++i)
		{
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i][j] = (i == j ?  1 : 0);
			};
		};
	}
};
public class TranslationMatrix : Matrix {
	public TranslationMatrix(double x, double y, double z) {
		_fieldGrid[0][0] = 1;
		_fieldGrid[0][3] = x;
		_fieldGrid[1][3] = y;
		_fieldGrid[1][1] = 1;
		_fieldGrid[2][3] = z;
		_fieldGrid[2][2] = 1;
		_fieldGrid[3][3] = 1;
	}

};
public class ScalingMatrix : Matrix {
	public ScalingMatrix(double x, double y, double z) {
		_fieldGrid[0][0] = x;
		_fieldGrid[1][1] = y;
		_fieldGrid[2][2] = z;
		_fieldGrid[3][3] = 1;
	}
};
public class XRotationMatrix : Matrix {
	public XRotationMatrix(double radians) {
		_fieldGrid[0][0] = 1;
		_fieldGrid[1][1] = Math.Cos(radians);
		_fieldGrid[1][2] = -Math.Sin(radians);
		_fieldGrid[2][1] = Math.Sin(radians);
		_fieldGrid[2][2] = Math.Cos(radians);
		_fieldGrid[3][3] = 1;
	}
};
public class YRotationMatrix : Matrix {
	public YRotationMatrix(double radians) {
		_fieldGrid[0][0] = Math.Cos(radians);
		_fieldGrid[0][2] = Math.Sin(radians);
		_fieldGrid[1][1] = 1;
		_fieldGrid[2][0] = -Math.Sin(radians);
		_fieldGrid[2][2] = Math.Cos(radians);
		_fieldGrid[3][3] = 1;
	}
};
public class ZRotationMatrix : Matrix {
	public ZRotationMatrix(double radians) {
		_fieldGrid[0][0] = Math.Cos(radians);
		_fieldGrid[0][1] = -Math.Sin(radians);
		_fieldGrid[1][0] = Math.Sin(radians);
		_fieldGrid[1][1] = Math.Cos(radians);
		_fieldGrid[2][2] = 1;
		_fieldGrid[3][3] = 1;
	}
};
public class ShearingMatrix : Matrix {
	public ShearingMatrix(double xy, double xz, double yx, double yz, double zx, double zy) {
		_fieldGrid[0][0] = 1;
		_fieldGrid[0][1] = xy;
		_fieldGrid[0][2] = xz;
		_fieldGrid[1][0] = yx;
		_fieldGrid[1][1] = 1;
		_fieldGrid[1][2] = yz;
		_fieldGrid[2][0] = zx;
		_fieldGrid[2][1] = zy;
		_fieldGrid[2][2] = 1;
		_fieldGrid[3][3] = 1;
	}
};
public class ViewMatrix : Matrix {
	public ViewMatrix()
	{
		for (int i = 0; i < _fieldRows; ++i)
		{
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i][j] = (i == j ?  1 : 0);
			};
		};
	}
	public ViewMatrix(Point argStart, Point argEnd, Vector argUp) {
		Vector varForward = (argEnd - argStart).GetNormal();
		Vector varLeft = varForward.GetCrossProduct(argUp.GetNormal());
		Vector varUp = varLeft.GetCrossProduct(varForward);
		List<double> varValues = new List<double>{varLeft._fieldX, varLeft._fieldY, varLeft._fieldZ, 0, varUp._fieldX, varUp._fieldY, varUp._fieldZ, 0, -varForward._fieldX, -varForward._fieldY, -varForward._fieldZ, 0, 0, 0, 0, 1};
		Matrix varOrientation = new Matrix(4,4, varValues);
		Matrix varResult = varOrientation * new TranslationMatrix(-argStart._fieldX, -argStart._fieldY, -argStart._fieldZ);
		for (int i = 0; i < _fieldRows; ++i)
		{
			for (int j = 0; j < _fieldColumns; ++j)
			{
				_fieldGrid[i][j] = varResult._fieldGrid[i][j];
			};
		};
	}
};